devolutions-gateway/src/token.rs

```diff
@@ -511,6 +536,63 @@ impl<'de> de::Deserialize<'de> for JmuxTokenClaims {
     }
 }
 
+// ----- jrec claims ----- //
+
+#[derive(Clone)]
+pub struct JrecTokenClaims {
+    /// Association ID (= Session ID)
+    pub jet_aid: Uuid,
+
+    /// Application protocol
+    pub jet_ap: ApplicationProtocol,
+
+    // Recording file type
+    pub jet_rft: RecordingFileType,
+
+    /// Max session duration
+    pub jet_ttl: SessionTtl,
+
+    // JWT expiration time claim.
+    // We need this to build our token invalidation cache.
+    // This doesn't need to be explicitly written in the structure to be checked by the JwtValidator.
+    exp: i64,
+
+    // Unique ID for this token
+    jti: Uuid,
+}
+
+impl<'de> de::Deserialize<'de> for JrecTokenClaims {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        struct ClaimsHelper {
+            #[serde(default = "Uuid::new_v4")]
+            jet_aid: Uuid,
+            jet_ap: ApplicationProtocol,
+            jet_rft: RecordingFileType,
+            #[serde(default)]
+            jet_ttl: u64,
+            exp: i64,
+            jti: Uuid,
+        }
+
+        let claims = ClaimsHelper::deserialize(deserializer)?;
+
+        let jet_ttl = SessionTtl::from(claims.jet_ttl);
+
+        Ok(Self {
+            jet_aid: claims.jet_aid,
+            jet_ap: claims.jet_ap,
+            jet_rft: claims.jet_rft,
+            jet_ttl,
+            exp: claims.exp,
+            jti: claims.jti,
+        })
+    }
+}
+
```

Comment:
In this case, you can derive `Deserialize` automatically.

Reply:
I'll ask you how to do it next time, it's not much of an issue to have a bit too much boilerplate for the current PR